### PR TITLE
Content length in stream #347

### PR DESCRIPTION
**Pull request by AWAS666**
This should fix the issue in streams.

Not sure about regular requests because of this:
https://github.com/aspnet/AspNetWebStack/issues/252#issuecomment-584305667

Might have to implement it differently to do that, which is why I'd preferably wait for feedback.

### CODE REVIEW

**Comment by kayhantolga** [comment]
Hi, thank you for the PR. Have you tested your changes? I couldn't see the content length header after testing your implementation. 🤔

**Comment by AWAS666** [comment]
I've tested it with the server mentioned in the issue #347 and that works just fine now.

**Comment by kayhantolga** [comment]
Interesting. I couldn't see any difference in the request when using the non-stream post method. Maybe I should test with the library you suggested.
Also, I have a question: I noticed that you only set the content-length for the stream API. Is that intentional?

**Comment by AWAS666** [comment]
The header isn't set in the request header but in the header of the content instead.

I changed it for non stream in the second commit, that one required more changes which is I split them up.

File: OpenAI.SDK/Extensions/HttpclientExtensions.cs [hunk 1]
```diff
@@ -28,6 +30,8 @@ public static HttpResponseMessage PostAsStreamAsync(this HttpClient client, stri
         using var request = new HttpRequestMessage(HttpMethod.Post, uri);
         request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));
         request.Content = content;
+        var jsonString = JsonSerializer.Serialize(requestModel,settings);
+        request.Content.Headers.ContentLength = Encoding.UTF8.GetByteCount(jsonString);
```
Comment (AWAS666) [inline OpenAI.SDK/Extensions/HttpclientExtensions.cs:37]: the content here has yet another header and there the length doesn't get set automatically for whatever reason.

**Comment by kayhantolga** [comment]
Hi, I was able to test the library against [[text-generation-webui/openai](https://github.com/oobabooga/text-generation-webui/blob/main/extensions/openai/README.md)](https://github.com/oobabooga/text-generation-webui/blob/main/extensions/openai/README.md) in both the dev branch and your branch.

The non-stream mode works in both your branch and the dev branch, so I don't think we need to make any changes there. You can see this in the screenshot.

With stream, it doesn't work in the dev branch, but it works in your branch. However, I have two concerns. First, we are serializing the object twice, which will decrease overall performance. Second, manually calculating content length is not recommended. The Httpclient library and the server should handle this between them. Incorrect calculations may cause unintended issues. I suggest creating an issue ticket in their library.

For now, our library can work with their library without stream support. I will close this PR and issue without merging.

Thank you for reporting this and attempting to find a solution.
image
image

**Pull request closed without merging**